File: js-port/funcscript-js/src/funcs/assert/assert-functions.js
const { BaseFunction, CallType } = require('../../core/function-base');
const helpers = require('../helpers');
const { FSDataType } = require('../../core/fstypes');

function ensureTyped(value) {
  return helpers.ensureTyped(value);
}

function formatValue(typed) {
  const type = helpers.typeOf(typed);
  const value = helpers.valueOf(typed);
  switch (type) {
    case FSDataType.Null:
      return 'null';
    case FSDataType.Boolean:
      return value ? 'true' : 'false';
    case FSDataType.Integer:
    case FSDataType.Float:
      return String(value);
    case FSDataType.BigInteger:
      return value.toString();
    case FSDataType.String:
      return JSON.stringify(value);
    case FSDataType.Error: {
      const err = value || {};
      return `${err.errorType || 'Error'}: ${err.errorMessage || ''}`;
    }
    case FSDataType.List:
      return '[list]';
    case FSDataType.KeyValueCollection:
      return '[object]';
    default:
      return String(value);
  }
}

function makeAssertionError(symbol, message) {
  return helpers.makeError('assert', `${symbol}: ${message}`);
}

function expectBoolean(symbol, typed) {
  const type = helpers.typeOf(typed);
  if (type !== FSDataType.Boolean) {
    return { ok: false, error: makeAssertionError(symbol, 'boolean expected') };
  }
  return { ok: true, value: helpers.valueOf(typed) === true };
}

function toNumber(symbol, value, label) {
  const typed = ensureTyped(value);
  const type = helpers.typeOf(typed);
  const raw = helpers.valueOf(typed);
  switch (type) {
    case FSDataType.Integer:
    case FSDataType.Float:
      return { ok: true, value: Number(raw) };
    case FSDataType.BigInteger:
      return { ok: true, value: Number(raw) };
    default:
      return { ok: false, error: makeAssertionError(symbol, `${label} must be numeric`) };
  }
}

function compareValues(left, right, symbol) {
  const comparison = helpers.compare(left, right, symbol);
  if (typeof comparison === 'number') {
    return { ok: true, value: comparison };
  }
  if (Array.isArray(comparison) && helpers.typeOf(comparison) === FSDataType.Error) {
    return { ok: false, error: comparison };
  }
  return { ok: false, error: makeAssertionError(symbol, 'Unable to compare values') };
}

function ensureErrorValue(symbol, typedValue) {
  if (helpers.typeOf(typedValue) !== FSDataType.Error) {
    return null;
  }
  return helpers.valueOf(typedValue) || {};
}

class AssertEqualFunction extends BaseFunction {
  constructor() {
    super();
    this.symbol = 'assert.equal';
    this.callType = CallType.Prefix;
  }

  get maxParameters() {
    return 2;
  }

  evaluate(provider, parameters) {
    if (parameters.count !== 2) {
      return makeAssertionError(this.symbol, 'Expected two parameters');
    }
    const left = ensureTyped(parameters.getParameter(provider, 0));
    const right = ensureTyped(parameters.getParameter(provider, 1));
    const result = compareValues(left, right, this.symbol);
    if (!result.ok) {
      return result.error;
    }
    if (result.value === 0) {
      return helpers.makeValue(FSDataType.Boolean, true);
    }
    return makeAssertionError(this.symbol, `${formatValue(left)} != ${formatValue(right)}`);
  }

  parName(index) {
    return index === 0 ? 'left' : 'right';
  }
}

class AssertNotEqualFunction extends BaseFunction {
  constructor() {
    super();
    this.symbol = 'assert.notEqual';
    this.callType = CallType.Prefix;
  }

  get maxParameters() {
    return 2;
  }

  evaluate(provider, parameters) {
    if (parameters.count !== 2) {
      return makeAssertionError(this.symbol, 'Expected two parameters');
    }
    const left = ensureTyped(parameters.getParameter(provider, 0));
    const right = ensureTyped(parameters.getParameter(provider, 1));
    const result = compareValues(left, right, this.symbol);
    if (!result.ok) {
      return result.error;
    }
    if (result.value !== 0) {
      return helpers.makeValue(FSDataType.Boolean, true);
    }
    return makeAssertionError(this.symbol, `${formatValue(left)} == ${formatValue(right)}`);
  }

  parName(index) {
    return index === 0 ? 'left' : 'right';
  }
}

class AssertGreaterFunction extends BaseFunction {
  constructor() {
    super();
    this.symbol = 'assert.greater';
    this.callType = CallType.Prefix;
  }

  get maxParameters() {
    return 2;
  }

  evaluate(provider, parameters) {
    if (parameters.count !== 2) {
      return makeAssertionError(this.symbol, 'Expected two parameters');
    }
    const left = ensureTyped(parameters.getParameter(provider, 0));
    const right = ensureTyped(parameters.getParameter(provider, 1));
    const result = compareValues(left, right, this.symbol);
    if (!result.ok) {
      return result.error;
    }
    if (result.value > 0) {
      return helpers.makeValue(FSDataType.Boolean, true);
    }
    return makeAssertionError(this.symbol, `Expected ${formatValue(left)} > ${formatValue(right)}`);
  }

  parName(index) {
    return index === 0 ? 'left' : 'right';
  }
}

class AssertLessFunction extends BaseFunction {
  constructor() {
    super();
    this.symbol = 'assert.less';
    this.callType = CallType.Prefix;
  }

  get maxParameters() {
    return 2;
  }

  evaluate(provider, parameters) {
    if (parameters.count !== 2) {
      return makeAssertionError(this.symbol, 'Expected two parameters');
    }
    const left = ensureTyped(parameters.getParameter(provider, 0));
    const right = ensureTyped(parameters.getParameter(provider, 1));
    const result = compareValues(left, right, this.symbol);
    if (!result.ok) {
      return result.error;
    }
    if (result.value < 0) {
      return helpers.makeValue(FSDataType.Boolean, true);
    }
    return makeAssertionError(this.symbol, `Expected ${formatValue(left)} < ${formatValue(right)}`);
  }

  parName(index) {
    return index === 0 ? 'left' : 'right';
  }
}

class AssertTrueFunction extends BaseFunction {
  constructor() {
    super();
    this.symbol = 'assert.true';
    this.callType = CallType.Prefix;
  }

  get maxParameters() {
    return 1;
  }

  evaluate(provider, parameters) {
    const error = helpers.expectParamCount(this.symbol, parameters, 1);
    if (error) {
      return error;
    }

    const typed = ensureTyped(parameters.getParameter(provider, 0));
    const boolResult = expectBoolean(this.symbol, typed);
    if (!boolResult.ok) {
      return boolResult.error;
    }
    if (boolResult.value) {
      return helpers.makeValue(FSDataType.Boolean, true);
    }
    return makeAssertionError(this.symbol, `Expected true but received ${formatValue(typed)}`);
  }
}

class AssertFalseFunction extends BaseFunction {
  constructor() {
    super();
    this.symbol = 'assert.false';
    this.callType = CallType.Prefix;
  }

  get maxParameters() {
    return 1;
  }

  evaluate(provider, parameters) {
    const error = helpers.expectParamCount(this.symbol, parameters, 1);
    if (error) {
      return error;
    }

    const typed = ensureTyped(parameters.getParameter(provider, 0));
    const boolResult = expectBoolean(this.symbol, typed);
    if (!boolResult.ok) {
      return boolResult.error;
    }
    if (!boolResult.value) {
      return helpers.makeValue(FSDataType.Boolean, true);
    }
    return makeAssertionError(this.symbol, `Expected false but received ${formatValue(typed)}`);
  }
}

class AssertApproxFunction extends BaseFunction {
  constructor() {
    super();
    this.symbol = 'assert.approx';
    this.callType = CallType.Prefix;
  }

  get maxParameters() {
    return 3;
  }

  evaluate(provider, parameters) {
    if (parameters.count !== 3) {
      return makeAssertionError(this.symbol, 'Expected three parameters');
    }
    const left = ensureTyped(parameters.getParameter(provider, 0));
    const right = ensureTyped(parameters.getParameter(provider, 1));
    const epsilonTyped = ensureTyped(parameters.getParameter(provider, 2));

    const leftNumber = toNumber(this.symbol, left, 'left');
    if (!leftNumber.ok) {
      return leftNumber.error;
    }
    const rightNumber = toNumber(this.symbol, right, 'right');
    if (!rightNumber.ok) {
      return rightNumber.error;
    }
    const epsilonNumber = toNumber(this.symbol, epsilonTyped, 'epsilon');
    if (!epsilonNumber.ok) {
      return epsilonNumber.error;
    }

    const distance = Math.abs(leftNumber.value - rightNumber.value);
    const limit = Math.abs(epsilonNumber.value);
    if (distance <= limit) {
      return helpers.makeValue(FSDataType.Boolean, true);
    }
    return makeAssertionError(
      this.symbol,
      `|${leftNumber.value} - ${rightNumber.value}| = ${distance} > ${limit}`
    );
  }
}

class AssertNoErrorFunction extends BaseFunction {
  constructor() {
    super();
    this.symbol = 'assert.noerror';
    this.callType = CallType.Prefix;
  }

  get maxParameters() {
    return 1;
  }

  evaluate(provider, parameters) {
    const error = helpers.expectParamCount(this.symbol, parameters, 1);
    if (error) {
      return error;
    }
    const typed = ensureTyped(parameters.getParameter(provider, 0));
    if (helpers.typeOf(typed) !== FSDataType.Error) {
      return helpers.makeValue(FSDataType.Boolean, true);
    }
    const err = helpers.valueOf(typed) || {};
    return makeAssertionError(
      this.symbol,
      `Expected non-error result but received ${err.errorType || 'Error'}: ${err.errorMessage || ''}`
    );
  }
}

class AssertIsErrorFunction extends BaseFunction {
  constructor() {
    super();
    this.symbol = 'assert.iserror';
    this.callType = CallType.Prefix;
  }

  get maxParameters() {
    return 1;
  }

  evaluate(provider, parameters) {
    const error = helpers.expectParamCount(this.symbol, parameters, 1);
    if (error) {
      return error;
    }
    const typed = ensureTyped(parameters.getParameter(provider, 0));
    if (helpers.typeOf(typed) === FSDataType.Error) {
      return helpers.makeValue(FSDataType.Boolean, true);
    }
    return makeAssertionError(this.symbol, 'Expected an error result but value was not an error.');
  }
}

class AssertIsErrorTypeFunction extends BaseFunction {
  constructor() {
    super();
    this.symbol = 'assert.iserrortype';
    this.callType = CallType.Prefix;
  }

  get maxParameters() {
    return 2;
  }

  evaluate(provider, parameters) {
    if (parameters.count !== 2) {
      return makeAssertionError(this.symbol, 'Expected value and type name');
    }
    const errorValue = ensureTyped(parameters.getParameter(provider, 0));
    const typeValue = helpers.requireString(this.symbol, parameters.getParameter(provider, 1), 'type');
    if (!typeValue.ok) {
      return typeValue.error;
    }

    const err = ensureErrorValue(this.symbol, errorValue);
    if (!err) {
      return makeAssertionError(this.symbol, 'Value is not an error result.');
    }
    const expected = typeValue.value.toLowerCase();
    const actual = (err.errorType || '').toLowerCase();
    if (actual === expected) {
      return helpers.makeValue(FSDataType.Boolean, true);
    }
    return makeAssertionError(
      this.symbol,
      `Expected error type ${expected || '(blank)'} but received ${actual || '(blank)'}`
    );
  }
}

class AssertHasErrorMessageFunction extends BaseFunction {
  constructor() {
    super();
    this.symbol = 'assert.hasErrorMessage';
    this.callType = CallType.Prefix;
  }

  get maxParameters() {
    return 2;
  }

  evaluate(provider, parameters) {
    if (parameters.count !== 2) {
      return makeAssertionError(this.symbol, 'Expected value and message');
    }
    const errorValue = ensureTyped(parameters.getParameter(provider, 0));
    const messageValue = helpers.requireString(this.symbol, parameters.getParameter(provider, 1), 'message');
    if (!messageValue.ok) {
      return messageValue.error;
    }

    const err = ensureErrorValue(this.symbol, errorValue);
    if (!err) {
      return makeAssertionError(this.symbol, 'Value is not an error result.');
    }

    const expected = messageValue.value;
    const actual = err.errorMessage || '';
    const result = expected ? actual.includes(expected) : actual.length === 0;
    if (result) {
      return helpers.makeValue(FSDataType.Boolean, true);
    }
    return makeAssertionError(
      this.symbol,
      `Expected error message containing ${JSON.stringify(expected)} but received ${JSON.stringify(actual)}`
    );
  }
}

class AssertIsNullFunction extends BaseFunction {
  constructor() {
    super();
    this.symbol = 'assert.isnull';
    this.callType = CallType.Prefix;
  }

  get maxParameters() {
    return 1;
  }

  evaluate(provider, parameters) {
    const error = helpers.expectParamCount(this.symbol, parameters, 1);
    if (error) {
      return error;
    }
    const typed = ensureTyped(parameters.getParameter(provider, 0));
    if (helpers.typeOf(typed) === FSDataType.Null) {
      return helpers.makeValue(FSDataType.Boolean, true);
    }
    return makeAssertionError(this.symbol, `Expected null but received ${formatValue(typed)}`);
  }
}

class AssertIsNotNullFunction extends BaseFunction {
  constructor() {
    super();
    this.symbol = 'assert.isnotnull';
    this.callType = CallType.Prefix;
  }

  get maxParameters() {
    return 1;
  }

  evaluate(provider, parameters) {
    const error = helpers.expectParamCount(this.symbol, parameters, 1);
    if (error) {
      return error;
    }
    const typed = ensureTyped(parameters.getParameter(provider, 0));
    if (helpers.typeOf(typed) !== FSDataType.Null) {
      return helpers.makeValue(FSDataType.Boolean, true);
    }
    return makeAssertionError(this.symbol, 'Expected non-null value but received null.');
  }
}

module.exports = {
  AssertEqualFunction,
  AssertNotEqualFunction,
  AssertGreaterFunction,
  AssertLessFunction,
  AssertTrueFunction,
  AssertFalseFunction,
  AssertApproxFunction,
  AssertNoErrorFunction,
  AssertIsErrorFunction,
  AssertIsErrorTypeFunction,
  AssertHasErrorMessageFunction,
  AssertIsNullFunction,
  AssertIsNotNullFunction
};

File: FuncScript/Functions/Test/AssertFunctions.cs
using FuncScript.Core;
using FuncScript.Model;
using System;
using System.Globalization;
using System.Numerics;

namespace FuncScript.Functions.Test
{
    internal static class AssertHelpers
    {
        public static double ToDouble(string symbol, object value, string parameterName)
        {
            if (value == null)
                throw new Error.TypeMismatchError($"{symbol}: {parameterName} must be numeric");

            return value switch
            {
                int i => i,
                long l => l,
                float f => f,
                double d => d,
                decimal dec => (double)dec,
                BigInteger big => (double)big,
                _ => throw new Error.TypeMismatchError($"{symbol}: {parameterName} must be numeric")
            };
        }

        public static string FormatValue(object value)
        {
            if (value == null)
                return "null";

            return value switch
            {
                string s => $"\"{s}\"",
                bool b => b.ToString().ToLowerInvariant(),
                IFormattable formattable => formattable.ToString(null, CultureInfo.InvariantCulture),
                _ => value.ToString() ?? value.GetType().Name
            };
        }

        public static bool IsError(object value, out FsError error)
        {
            if (value is FsError fsError)
            {
                error = fsError;
                return true;
            }

            error = null;
            return false;
        }

        public static bool AreEqual(object left, object right)
        {
            if (left == null && right == null)
                return true;

            if (left == null || right == null)
                return false;

            if (Engine.IsNumeric(left) && Engine.IsNumeric(right))
                Engine.ConvertToCommonNumericType(left, right, out left, out right);

            return left?.GetType() == right?.GetType() && left.Equals(right);
        }

        public static bool TryCompare(object left, object right, string symbol, out int comparison, out FsError error)
        {
            comparison = 0;
            error = null;

            if (left == null || right == null)
            {
                error = new FsError(FsError.ERROR_TYPE_MISMATCH, $"{symbol}: values cannot be null");
                return false;
            }

            if (Engine.IsNumeric(left) && Engine.IsNumeric(right))
                Engine.ConvertToCommonNumericType(left, right, out left, out right);

            if (left.GetType() != right.GetType())
            {
                error = new FsError(FsError.ERROR_TYPE_MISMATCH, $"{symbol}: incompatible types");
                return false;
            }

            if (left is IComparable comparable)
            {
                comparison = comparable.CompareTo(right);
                return true;
            }

            error = new FsError(FsError.ERROR_TYPE_MISMATCH, $"{symbol}: values are not comparable");
            return false;
        }
    }

    public abstract class AssertionFunctionBase : IFsFunction
    {
        public abstract string Symbol { get; }

        public virtual int MaxParsCount => 1;
        public virtual CallType CallType => CallType.Prefix;
        public virtual int Precedence => 0;

        public abstract object Evaluate(object par);

        protected FsList ExpectParameters(object par)
            => FunctionArgumentHelper.ExpectList(par, this.Symbol);

        protected FsError Failure(string message)
            => new FsError(FsError.ERROR_DEFAULT, $"{Symbol}: {message}");

        public virtual string ParName(int index) => string.Empty;
    }

    #region Equality and Comparison

    [ProviderCollection("assert", MemberNames = new[] { "equal" })]
    public sealed class AssertEqualFunction : AssertionFunctionBase
    {
        public override string Symbol => "assert.equal";

        public override int MaxParsCount => 2;

        public override object Evaluate(object par)
        {
            var pars = ExpectParameters(par);
            if (pars.Length != 2)
                throw new Error.TypeMismatchError($"{Symbol}: expected two parameters");

            var left = pars[0];
            var right = pars[1];

            if (AssertHelpers.AreEqual(left, right))
                return true;

            return Failure($"{AssertHelpers.FormatValue(left)} != {AssertHelpers.FormatValue(right)}");
        }

        public override string ParName(int index) => index == 0 ? "left" : "right";
    }

    [ProviderCollection("assert", MemberNames = new[] { "notEqual" })]
    public sealed class AssertNotEqualFunction : AssertionFunctionBase
    {
        public override string Symbol => "assert.notEqual";

        public override int MaxParsCount => 2;

        public override object Evaluate(object par)
        {
            var pars = ExpectParameters(par);
            if (pars.Length != 2)
                throw new Error.TypeMismatchError($"{Symbol}: expected two parameters");

            var left = pars[0];
            var right = pars[1];

            if (AssertHelpers.AreEqual(left, right))
                return Failure("Values should not be equal but they are.");

            return true;
        }

        public override string ParName(int index) => index == 0 ? "left" : "right";
    }

    [ProviderCollection("assert", MemberNames = new[] { "greater" })]
    public sealed class AssertGreaterFunction : AssertionFunctionBase
    {
        public override string Symbol => "assert.greater";

        public override int MaxParsCount => 2;

        public override object Evaluate(object par)
        {
            var pars = ExpectParameters(par);
            if (pars.Length != 2)
                throw new Error.TypeMismatchError($"{Symbol}: expected two parameters");

            var left = pars[0];
            var right = pars[1];

            if (!AssertHelpers.TryCompare(left, right, Symbol, out var cmp, out var error))
                return error;

            if (cmp > 0)
                return true;

            return Failure($"Expected {AssertHelpers.FormatValue(left)} > {AssertHelpers.FormatValue(right)}");
        }

        public override string ParName(int index) => index == 0 ? "left" : "right";
    }

    [ProviderCollection("assert", MemberNames = new[] { "less" })]
    public sealed class AssertLessFunction : AssertionFunctionBase
    {
        public override string Symbol => "assert.less";

        public override int MaxParsCount => 2;

        public override object Evaluate(object par)
        {
            var pars = ExpectParameters(par);
            if (pars.Length != 2)
                throw new Error.TypeMismatchError($"{Symbol}: expected two parameters");

            var left = pars[0];
            var right = pars[1];

            if (!AssertHelpers.TryCompare(left, right, Symbol, out var cmp, out var error))
                return error;

            if (cmp < 0)
                return true;

            return Failure($"Expected {AssertHelpers.FormatValue(left)} < {AssertHelpers.FormatValue(right)}");
        }

        public override string ParName(int index) => index == 0 ? "left" : "right";
    }

    #endregion

    [ProviderCollection("assert", MemberNames = new[] { "true" })]
    public sealed class AssertTrueFunction : AssertionFunctionBase
    {
        public override string Symbol => "assert.true";

        public override object Evaluate(object par)
        {
            var pars = ExpectParameters(par);
            if (pars.Length != 1)
                throw new Error.TypeMismatchError($"{Symbol}: boolean argument expected");

            var value = pars[0];
            if (value is bool boolValue)
                return boolValue ? true : Failure($"Expected true but was {AssertHelpers.FormatValue(value)}");

            throw new Error.TypeMismatchError($"{Symbol}: boolean argument expected");
        }

        public override string ParName(int index) => "value";
    }

    [ProviderCollection("assert", MemberNames = new[] { "false" })]
    public sealed class AssertFalseFunction : AssertionFunctionBase
    {
        public override string Symbol => "assert.false";

        public override object Evaluate(object par)
        {
            var pars = ExpectParameters(par);
            if (pars.Length != 1)
                throw new Error.TypeMismatchError($"{Symbol}: boolean argument expected");

            var value = pars[0];
            if (value is bool boolValue)
                return !boolValue ? true : Failure($"Expected false but was {AssertHelpers.FormatValue(value)}");

            throw new Error.TypeMismatchError($"{Symbol}: boolean argument expected");
        }

        public override string ParName(int index) => "value";
    }

    [ProviderCollection("assert", MemberNames = new[] { "approx" })]
    public sealed class AssertApproxFunction : AssertionFunctionBase
    {
        public override string Symbol => "assert.approx";

        public override int MaxParsCount => 3;

        public override object Evaluate(object par)
        {
            var pars = ExpectParameters(par);
            if (pars.Length != 3)
                throw new Error.TypeMismatchError($"{Symbol}: expected 3 parameters");

            var left = AssertHelpers.ToDouble(Symbol, pars[0], "left");
            var right = AssertHelpers.ToDouble(Symbol, pars[1], "right");
            var epsilon = System.Math.Abs(AssertHelpers.ToDouble(Symbol, pars[2], "epsilon"));
            var difference = System.Math.Abs(left - right);

            if (difference <= epsilon)
                return true;

            return Failure($"|{left} - {right}| = {difference} > {epsilon}");
        }

        public override string ParName(int index) => index switch
        {
            0 => "left",
            1 => "right",
            2 => "epsilon",
            _ => string.Empty
        };
    }

    [ProviderCollection("assert", MemberNames = new[] { "noerror" })]
    public sealed class AssertNoErrorFunction : AssertionFunctionBase
    {
        public override string Symbol => "assert.noerror";

        public override object Evaluate(object par)
        {
            var pars = ExpectParameters(par);
            if (pars.Length != 1)
                throw new Error.TypeMismatchError($"{Symbol}: argument expected");

            var value = pars[0];
            if (value is FsError error)
                return Failure($"Expected non-error result but received {error.ErrorType}: {error.ErrorMessage}");

            return true;
        }

        public override string ParName(int index) => "value";
    }

    [ProviderCollection("assert", MemberNames = new[] { "iserror" })]
    public sealed class AssertIsErrorFunction : AssertionFunctionBase
    {
        public override string Symbol => "assert.iserror";

        public override object Evaluate(object par)
        {
            var pars = ExpectParameters(par);
            if (pars.Length != 1)
                throw new Error.TypeMismatchError($"{Symbol}: argument expected");

            if (pars[0] is FsError)
                return true;

            return Failure("Expected an error result but received a non-error value.");
        }

        public override string ParName(int index) => "value";
    }

    [ProviderCollection("assert", MemberNames = new[] { "iserrortype" })]
    public sealed class AssertIsErrorTypeFunction : AssertionFunctionBase
    {
        public override string Symbol => "assert.iserrortype";

        public override int MaxParsCount => 2;

        public override object Evaluate(object par)
        {
            var pars = ExpectParameters(par);
            if (pars.Length != 2)
                throw new Error.TypeMismatchError($"{Symbol}: expected value and type name");

            var first = pars[0];
            var typeName = pars[1] as string
                ?? throw new Error.TypeMismatchError($"{Symbol}: type name must be a string");

            if (!AssertHelpers.IsError(first, out var error))
                return Failure("Value is not an error result.");

            if (string.Equals(error?.ErrorType, typeName, StringComparison.OrdinalIgnoreCase))
                return true;

            return Failure($"Expected error type {typeName} but received {error?.ErrorType ?? "(blank)"}.");
        }

        public override string ParName(int index) => index == 0 ? "value" : "type";
    }

    [ProviderCollection("assert", MemberNames = new[] { "haserrormessage" })]
    public sealed class AssertHasErrorMessageFunction : AssertionFunctionBase
    {
        public override string Symbol => "assert.haserrormessage";

        public override int MaxParsCount => 2;

        public override object Evaluate(object par)
        {
            var pars = ExpectParameters(par);
            if (pars.Length != 2)
                throw new Error.TypeMismatchError($"{Symbol}: expected value and message");

            var first = pars[0];
            var message = pars[1] as string
                ?? throw new Error.TypeMismatchError($"{Symbol}: message must be a string");

            if (!AssertHelpers.IsError(first, out var error))
                return Failure("Value is not an error result.");

            var actual = error?.ErrorMessage ?? string.Empty;
            if (string.IsNullOrEmpty(message))
                return string.IsNullOrEmpty(actual) ? true : Failure($"Expected empty error message but received {actual}");

            return actual.IndexOf(message, StringComparison.OrdinalIgnoreCase) >= 0
                ? true
                : Failure($"Expected error message containing '{message}' but received '{actual}'");
        }

        public override string ParName(int index) => index == 0 ? "value" : "message";
    }

    [ProviderCollection("assert", MemberNames = new[] { "isnull" })]
    public sealed class AssertIsNullFunction : AssertionFunctionBase
    {
        public override string Symbol => "assert.isnull";

        public override object Evaluate(object par)
        {
            var pars = ExpectParameters(par);
            if (pars.Length != 1)
                throw new Error.TypeMismatchError($"{Symbol}: argument expected");

            if (pars[0] == null)
                return true;

            return Failure("Expected null but received a non-null value.");
        }

        public override string ParName(int index) => "value";
    }

    [ProviderCollection("assert", MemberNames = new[] { "isnotnull" })]
    public sealed class AssertIsNotNullFunction : AssertionFunctionBase
    {
        public override string Symbol => "assert.isnotnull";

        public override object Evaluate(object par)
        {
            var pars = ExpectParameters(par);
            if (pars.Length != 1)
                throw new Error.TypeMismatchError($"{Symbol}: argument expected");

            if (pars[0] != null)
                return true;

            return Failure("Expected a non-null value but received null.");
        }

        public override string ParName(int index) => "value";
    }
}
